CHANGES.md

```diff
@@ -36,6 +36,15 @@ OpenSSL 3.4
 
    *Stephan Wurm*
 
+ * OpenSSL libraries no longer use atexit(3) to perform cleanup.
+   The cleanup now runs automatically via a library destructor.
+   Note that invocation of exit handlers registered by the application
+   via OPENSSL_atexit() is arranged by the linker. They are no longer
+   executed via atexit(3). If your application is linked statically,
+   then library destructors are never executed.
```

Comment:
If this is really a problem on Windows the existing code might actually be better in this regard because on Windows we use `_onexit` rather than `atexit`. Which is described like this:

````
The _onexit function is passed the address of a function (function) to be called when the program terminates normally. Successive calls to _onexit create a register of functions that are executed in LIFO (last-in-first-out) order. The functions passed to _onexit can't take parameters.

In the case when _onexit is called from within a DLL, routines registered with _onexit run when the DLL is unloaded, after DllMain is called with DLL_PROCESS_DETACH.
````
https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/onexit-onexit-m?view=msvc-170

Reply:
Hmm... you're right. I did not know that library destructors are run even if statically linked to an executable and not inside a shared library. So please @Sashan revert this part.

Reply:
Removing the use of `atexit()` is generally a good idea when there is a mix of static and dynamic linking of the same method. The registry is sometimes not cleaned during a DLL unload - it is unclear whether it should be from the POSIX spec. We have had may issues with this over the years in this project.

Reply:
this is my simple test case to see how things work:
```
lifty$ cat static.c                                                                                                                                                                                                                                                                                                        
#include <stdio.h>
#include <stdlib.h>

void
_fini(void)
{
        printf("%s\n", __func__);
}

static void
exit_handler(void)
{
        printf("%s\n", __func__);
}

int
main(int argc, const char* const * argv)
{
        atexit(exit_handler);
        return (0);
}
lifty$ cc -std=c89 -o dynamic static.c
lifty$ cc -std=c89 -static -o static static.c 
lifty$ ./dynamic                                                                                                                                                                                                                                                                                                           
exit_handler
_fini
lifty$ ./static                                                                                                                                                                                                                                                                                                            
exit_handler
lifty$ ldd static
static:
        Start            End              Type  Open Ref GrpRef Name
        00000a8c9877b000 00000a8c987a3000 dlib  1    0   0      /tmp/static
lifty$ ldd dynamic                                                                                                                                                                                                                                                                                                         
dynamic:
        Start            End              Type  Open Ref GrpRef Name
        00000e15cb36c000 00000e15cb370000 exe   1    0   0      dynamic
        00000e1822a3b000 00000e1822b32000 rlib  0    1   0      /usr/lib/libc.so.99.0
        00000e181e882000 00000e181e882000 ld.so 0    1   0      /usr/libexec/ld.so
```
may be my test is too simple.

Reply:
Yeah no-shared does not use `-static` to link the executable. The libc is still linked dynamically -> the destructor will be called. However `-static` links everything statically and the destructor won't be called.

Reply:
and now for linux:
```
#
# original version with _fini() fails to link with error
# static.c:(.text+0x0): multiple definition of `_fini'; /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/crti.o:(.fini+0x0): first defined here
# So I've opted for __attribute__ destructor variant
#
sashan@ubuntu:/tmp$ cat static.c 
#include <stdio.h>
#include <stdlib.h>

void __attribute__((destructor))
do_fini(void)
{
        printf("%s\n", __func__);
}

static void
exit_handler(void)
{
        printf("%s\n", __func__);
}

int
main(int argc, const char* const * argv)
{
        atexit(exit_handler);
        return (0);
}
sashan@ubuntu:/tmp$ cc -std=c89 -o dynamic  static.c
sashan@ubuntu:/tmp$ cc -std=c89 -static -o static  static.c
sashan@ubuntu:/tmp$ ./dynamic 
exit_handler
do_fini
sashan@ubuntu:/tmp$ ./static 
exit_handler
do_fini
sashan@ubuntu:/tmp$ ldd static
        not a dynamic executable
sashan@ubuntu:/tmp$ ldd dynamic
        linux-vdso.so.1 (0x00007ffe52fe6000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fbc06abd000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fbc06cf3000)
```

Reply:
after bringing the __attribute__ modification from Linux back to OpenBSD, things indeed work same for dynamic and static variant.
```
lifty$ cat static.c                                                                                                                                                                                                                                                                                                        
#include <stdio.h>
#include <stdlib.h>

#if 0
void
_fini(void)
{
        printf("%s\n", __func__);
}
#endif

void __attribute__((destructor))
do_fini(void)
{
        printf("%s\n", __func__);
}

static void
exit_handler(void)
{
        printf("%s\n", __func__);
}

int
main(int argc, const char* const * argv)
{
        atexit(exit_handler);
        return (0);
}
lifty$ cc -std=c89 -static -o static static.c
lifty$ cc -std=c89 -o dynamic static.c        
lifty$ ./static                                                                                                                                                                                                                                                                                                            
exit_handler
do_fini
lifty$ ./dynamic                                                                                                                                                                                                                                                                                                           
exit_handler
do_fini
lifty$ ldd static
static    static.c  
lifty$ ldd static 
static:
        Start            End              Type  Open Ref GrpRef Name
        00000b737169c000 00000b73716c4000 dlib  1    0   0      /tmp/static
lifty$ ldd dynamic
dynamic:
        Start            End              Type  Open Ref GrpRef Name
        00000cb3ff6e0000 00000cb3ff6e4000 exe   1    0   0      dynamic
        00000cb625140000 00000cb625237000 rlib  0    1   0      /usr/lib/libc.so.99.0
        00000cb69298e000 00000cb69298e000 ld.so 0    1   0      /usr/libexec/ld.so
```
so things are more colorful...

Reply:
I agree the statement in my comment is too strong. will delete it.